keps/sig-network/NNNN-kubelet-probes/README.md

```diff
@@ -0,0 +1,1039 @@
+
+    - (For pods that really have a strong use case for checking
+      connectivity in some non-standard way, such as via a load
+      balancer IP, it would still be possible for a user to write an
+      exec probe that uses `curl` or the like to connect to that IP
+      from the pod itself.)
+
+    - If we implement this by creating a new set of probe types,
+      rather than by modifying the semantics of the existing probes,
+      then it should be possible to block the old probe types via Pod
+      Security admission, so that administrators can create
+      AdminNetworkPolicies that don't include the kubelet probe hole.
+
+- Allow dual-stack / wrong-single-stack probing, either based on
+  information in the probe definition (e.g. `ipFamily: IPv6` or
+  `ipFamilyPolicy: RequireDualStack`) or by always doing "[Happy
+  Eyeballs]" (probing IPv4 and IPv6 in parallel and only requiring one
+  to succeed).
+
+- Make it easy for users who aren't doing "weird" things (eg,
+  readiness probe via LoadBalancer) to transition from the old probe
+  system to the new one.
+
+    - This might mean that they don't need to do anything, or it might
+      mean that there's an easy transition, like just changing
+      `tcpSocket:` to `tcp:` in their pod's probe definition.
+
+[`TCPSocketAction.Host`]: https://github.com/kubernetes/kubernetes/blob/v1.28.0/staging/src/k8s.io/api/core/v1/types.go#L2248
+[`HTTPGetAction.Host`]: https://github.com/kubernetes/kubernetes/blob/v1.28.0/staging/src/k8s.io/api/core/v1/types.go#L2219
+[Pod Security admission]: https://kubernetes.io/docs/concepts/security/pod-security-admission/
+[Happy Eyeballs]: https://tools.ietf.org/html/rfc8305
+
+### Non-Goals
+
+- Adding entirely-new probe types.
+
+- Changing the definition of NetworkPolicy v1 to remove the "kubelet
+  can access all local pods" hole, even in the case where the cluster
+  is configured in a way to forbid the use of probes that require the
+  hole. Even if _kubelet_ no longer needs to access pods, some users
+  likely depend on that rule to allow access to some pods from other
+  host-level processes, and there is no other way in NetworkPolicy v1
+  to write such a rule if we remove the built-in one. (However,
+  AdminNetworkPolicy and any future successor to NetworkPolicy should
+  be designed to not have this rule by default.)
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### API Changes to Fix the SSRF
+
+Regardless of what else we do, we will deprecate the
+`TCPSocketAction.Host` and `HTTPGetAction.Host` fields, and allow them
+to be blocked by Pod Security admission.
```

Comment:
I wonder why this is not the case today, cc @tallclair @liggitt

Reply:
related https://github.com/kubernetes/kubernetes/issues/81129

Reply:
Yeah, IMO we should just make this change. I think it is stretching the scope of what PSA covers a bit, but it's not that far.